kubeflow/application/prototypes/application.jsonnet

```diff
@@ -4,9 +4,9 @@
 // @shortDescription application Component
 // @param name string Name
 // @optionalParam type string kubeflow Type of application.
-// @optionalParam version string 0.3 Version of kubeflow
-// @optionalParam components array [] names of components or empty for all components
-// @optionalParam emitCRD string true Whether to emit the CustomResourceDefinition for Application
+// @optionalParam version string 0.4 Version of kubeflow
+// @optionalParam components array ["ambassador","jupyter","centraldashboard","tf-job-operator","pytorch-operator","spartakus","argo","pipeline"] names of components or empty for all components
```

Comment:
I thought we were going to use labels to match the Application CR with the resources that belong to that resource?

IIUC the code correctly we're relying on kfctl to dynamically build up the components value. This seems a little bit brittle (i.e. a bug in the code could easily lead us  to miss  setting the components).

Is this because we have a Kubeflow Application but we have no uber resource for each individual application just a bunch of individual K8s resources so it would be obvious looking at the Kubeflow Application resource what sub components were installed.

What if we did the following
Define an Application CR for Kubeflow

Define an Application CR for say Jupyter

Can we use labels to indicate that the Jupyter Application CR is owned by the Kubeflow Application CR and the Jupyter Application CR owns all the individual K8s resources associated with Jupyter?

Would this eliminate the need to rely on building up the components param via kfctl and just rely on adding the appropriate label consistently to link the Application CRs?

Reply:
> I thought we were going to use labels to match the Application CR with the resources that belong to that resource?

We are using labels. The Application CR has a spec.selector
```yaml
  selector:
    matchLabels:
      app.kubernetes.io/name: kubeflow
```
and each resource has a matching label

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: 2019-01-10T01:11:03Z
  generation: 5368
  labels:
    app: centraldashboard
    app.kubernetes.io/name: kubeflow
    ksonnet.io/component: application
```
This is per the Application Documentation.

> Is this because we have a Kubeflow Application but we have no uber resource for each individual application just a bunch of individual K8s resources so it would be obvious looking at the Kubeflow Application resource what sub components were installed.

yes. Although for Deployments, StatefulSets I'm not adding a label to the template. The application-controller claims ownership of the Deployment/StatefulSet so cascaded deletes work correctly, but I should probably add a label to all resource's spec.templates.

> What if we did the following
> Define an Application CR for Kubeflow
> 
> Define an Application CR for say Jupyter
> 
> Can we use labels to indicate that the Jupyter Application CR is owned by the Kubeflow Application CR and the Jupyter Application CR owns all the individual K8s resources associated with Jupyter?

That's an interesting idea. Let me do a bit of exploration on that. I wonder if a hierarchical name would work - eg

```yaml
  labels:
    app: centraldashboard
    app.kubernetes.io/name: kubeflow.jupyter
```

> Would this eliminate the need to rely on building up the components param via kfctl and just rely on adding the appropriate label consistently to link the Application CRs?

We could try and add a label to all resources in a given component via a parameter, but I think it would be very brittle. This solution admittedly does labeling dynamically by effectively doing a `map(addlabel, listofresources(listofcomponents)))` and appending the updated resources to the application's parts. It ends up being much faster to deploy and I take some care to ensure the component list is correct. One nice feature is you can deploy your own components by setting the env var DEFAULT_KUBEFLOW_COMPONENTS. EG 
```bash
export DEFAULT_KUBEFLOW_COMPONENTS='"ambassador","centraldashboard","tf-job-operator","tensorboard"'
```

I show this in the README Example.

Reply:
If its a quick fix it might be worth doing. 

I'd be ok with just opening up a P2 issue to track. I don't think its super high priority.

I think higher priority would be establishing a pattern where we have each package defining its own Application CR and then having all of these link up to the Kubeflow App. If its not a quick fix we can do it in a follow on.

I think the main reason to establish that pattern (e.g. doing it for a single application) is then it becomes easier to get others to contribute to doing it for other packages.

Reply:
> I think higher priority would be establishing a pattern where we have each package defining its own Application CR and then having all of these link up to the Kubeflow App. If its not a quick fix we can do it in a follow on.
> 
> I think the main reason to establish that pattern (e.g. doing it for a single application) is then it becomes easier to get others to contribute to doing it for other packages.

opened a separate issue #2257